# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [2]:
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

## Regression Model Evaluation

In [3]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
data = pd.read_csv('housing.csv', header=None, delimiter=r"\s+", names=column_names)

In [4]:
"""
CRIM - per capita crime rate by town
ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
INDUS - proportion of non-retail business acres per town.
CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
NOX - nitric oxides concentration (parts per 10 million)
RM - average number of rooms per dwelling
AGE - proportion of owner-occupied units built prior to 1940
DIS - weighted distances to five Boston employment centres
RAD - index of accessibility to radial highways
TAX - full-value property-tax rate per $10,000
PTRATIO - pupil-teacher ratio by town
B - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
LSTAT - % lower status of the population
MEDV - Median value of owner-occupied homes in $1000's"""

"\nCRIM - per capita crime rate by town\nZN - proportion of residential land zoned for lots over 25,000 sq.ft.\nINDUS - proportion of non-retail business acres per town.\nCHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)\nNOX - nitric oxides concentration (parts per 10 million)\nRM - average number of rooms per dwelling\nAGE - proportion of owner-occupied units built prior to 1940\nDIS - weighted distances to five Boston employment centres\nRAD - index of accessibility to radial highways\nTAX - full-value property-tax rate per $10,000\nPTRATIO - pupil-teacher ratio by town\nB - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town\nLSTAT - % lower status of the population\nMEDV - Median value of owner-occupied homes in $1000's"

In [5]:
#data

# Imports necesarios
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Definimos los nombres de las columnas tal como aparece en la consigna
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 
                'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']

# Cargamos el dataset desde el CSV (está en la misma carpeta)
data = pd.read_csv('housing.csv', header=None, delim_whitespace=True, names=column_names)

## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [6]:
# Your code here :
# ✅ 1. Separar los datos en entrenamiento (80%) y test (20%)

# Variable objetivo: MEDV (precio medio de las viviendas)
X = data.drop(columns='MEDV')
y = data['MEDV']

# División train-test (80-20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [7]:
# Your code here :
# ✅ 2. Entrenar un modelo de regresión lineal y hacer predicciones en train y test

# Crear el modelo
model = LinearRegression()

# Entrenar el modelo con los datos de entrenamiento
model.fit(X_train, y_train)

# Predicciones
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Mostrar algunas predicciones (opcional)
print("Predicciones en train:", y_train_pred[:5])
print("Predicciones en test:", y_test_pred[:5])


Predicciones en train: [10.96952405 19.41196567 23.06419602 12.1470648  18.3738116 ]
Predicciones en test: [28.99672362 36.02556534 14.81694405 25.03197915 18.76987992]


## 3. Calculate and print R-squared for both the training and the testing set.

In [8]:
# Your code here :
from sklearn.metrics import r2_score

# Calcular R² para entrenamiento y prueba
r2_train = r2_score(y_train, y_train_pred)
r2_test = r2_score(y_test, y_test_pred)

print(f"R² - Train: {r2_train:.2f}")
print(f"R² - Test: {r2_test:.2f}")


R² - Train: 0.75
R² - Test: 0.67


## 4. Calculate and print mean squared error for both the training and the testing set.

In [9]:
# Your code here :
from sklearn.metrics import mean_squared_error

# Calcular MSE para entrenamiento y prueba
mse_train = mean_squared_error(y_train, y_train_pred)
mse_test = mean_squared_error(y_test, y_test_pred)

print(f"Mean Squared Error - Train: {mse_train:.2f}")
print(f"Mean Squared Error - Test: {mse_test:.2f}")


Mean Squared Error - Train: 21.64
Mean Squared Error - Test: 24.29


## 5. Calculate and print mean absolute error for both the training and the testing set.

In [10]:
# Your code here :
from sklearn.metrics import mean_absolute_error

# Calcular MAE para entrenamiento y prueba
mae_train = mean_absolute_error(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)

print(f"Mean Absolute Error - Train: {mae_train:.2f}")
print(f"Mean Absolute Error - Test: {mae_test:.2f}")


Mean Absolute Error - Train: 3.31
Mean Absolute Error - Test: 3.19


## Classification Model Evaluation

In [11]:
from sklearn.datasets import load_iris
data = load_iris()

In [12]:
print(data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

In [13]:
column_names = data.feature_names

In [14]:
df = pd.DataFrame(data['data'],columns=column_names)

In [15]:
df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [16]:
target = pd.DataFrame(data.target)

In [17]:
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [18]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [19]:
# Your code here :
from sklearn.model_selection import train_test_split

# Variables predictoras y target
X = df
y = target

# Dividir en train y test (80% - 20%)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [20]:
# Your code here :
from sklearn.linear_model import LogisticRegression

# Crear y entrenar el modelo
model = LogisticRegression(max_iter=200)  # aumentamos max_iter para asegurar convergencia
model.fit(X_train, y_train.values.ravel())  # .ravel() para evitar warning de shape


LogisticRegression(max_iter=200)

## 8. Calculate and print the accuracy score for both the training and the testing set.

In [21]:
# Your code here :
from sklearn.metrics import accuracy_score

# Predicciones
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Calcular accuracy
train_accuracy = accuracy_score(y_train, y_train_pred)
test_accuracy = accuracy_score(y_test, y_test_pred)

print(f"Train Accuracy: {train_accuracy:.2f}")
print(f"Test Accuracy: {test_accuracy:.2f}")


Train Accuracy: 0.97
Test Accuracy: 1.00


## 9. This is placeholder for the AI. You don't need to do anything here.

In [22]:
# Your code here :

# Realizar predicciones
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

# Mostrar las primeras 5 predicciones de cada conjunto
print("Train Predictions (primeros 5):", y_train_pred[:5].ravel())
print("Test Predictions (primeros 5):", y_test_pred[:5].ravel())


Train Predictions (primeros 5): [0 0 1 0 0]
Test Predictions (primeros 5): [1 0 2 1 1]


## 10. Calculate and print the precision score for both the training and the testing set.

In [23]:
# Your code here :
from sklearn.metrics import precision_score

# Precision para entrenamiento y prueba
precision_train = precision_score(y_train, y_train_pred, average='macro')
precision_test = precision_score(y_test, y_test_pred, average='macro')

print(f"Precision Score - Train: {precision_train:.2f}")
print(f"Precision Score - Test: {precision_test:.2f}")


Precision Score - Train: 0.98
Precision Score - Test: 1.00


## 11. Calculate and print the recall score for both the training and the testing set.

In [24]:
# Your code here :
from sklearn.metrics import recall_score

# Recall para entrenamiento y prueba
recall_train = recall_score(y_train, y_train_pred, average='macro')
recall_test = recall_score(y_test, y_test_pred, average='macro')

print(f"Recall Score - Train: {recall_train:.2f}")
print(f"Recall Score - Test: {recall_test:.2f}")


Recall Score - Train: 0.98
Recall Score - Test: 1.00


## 12. Calculate and print the F1 score for both the training and the testing set.

In [25]:
# Your code here :
from sklearn.metrics import f1_score

# F1 score para entrenamiento y prueba
f1_train = f1_score(y_train, y_train_pred, average='macro')
f1_test = f1_score(y_test, y_test_pred, average='macro')

print(f"F1 Score - Train: {f1_train:.2f}")
print(f"F1 Score - Test: {f1_test:.2f}")


F1 Score - Train: 0.97
F1 Score - Test: 1.00


## 13. Generate confusion matrices for both the training and the testing set.

In [26]:
# Your code here :
from sklearn.metrics import confusion_matrix

# Matrices de confusión
conf_matrix_train = confusion_matrix(y_train, y_train_pred)
conf_matrix_test = confusion_matrix(y_test, y_test_pred)

print("Confusion Matrix - Train:")
print(conf_matrix_train)

print("\nConfusion Matrix - Test:")
print(conf_matrix_test)


Confusion Matrix - Train:
[[40  0  0]
 [ 0 38  3]
 [ 0  0 39]]

Confusion Matrix - Test:
[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

In [ ]:
# Have fun here !

# ✅ Propuesta: Comparar 3 modelos
# LogisticRegression
# KNeighborsClassifier
# RandomForestClassifier
# Usaremos accuracy, precision, recall y f1-score para comparar.



In [28]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Lista de modelos a evaluar
models = {
    "Logistic Regression": LogisticRegression(max_iter=200),
    "K-Nearest Neighbors": KNeighborsClassifier(n_neighbors=5),
    "Random Forest": RandomForestClassifier(n_estimators=100, random_state=42)
}

# Evaluar cada modelo
for name, model in models.items():
    model.fit(X_train, y_train.values.ravel())
    y_pred = model.predict(X_test)
    
    acc = accuracy_score(y_test, y_pred)
    prec = precision_score(y_test, y_pred, average='macro')
    rec = recall_score(y_test, y_pred, average='macro')
    f1 = f1_score(y_test, y_pred, average='macro')
    
    print(f"\n=== {name} ===")
    print(f"Accuracy : {acc:.2f}")
    print(f"Precision: {prec:.2f}")
    print(f"Recall   : {rec:.2f}")
    print(f"F1 Score : {f1:.2f}")



=== Logistic Regression ===
Accuracy : 1.00
Precision: 1.00
Recall   : 1.00
F1 Score : 1.00

=== K-Nearest Neighbors ===
Accuracy : 1.00
Precision: 1.00
Recall   : 1.00
F1 Score : 1.00

=== Random Forest ===
Accuracy : 1.00
Precision: 1.00
Recall   : 1.00
F1 Score : 1.00


📊 Qué verás:
Al ejecutar este bloque, obtendrás un resumen por modelo como este:

yaml
Copiar
Editar
=== Logistic Regression ===
Accuracy : 0.97
Precision: 0.97
Recall   : 0.97
F1 Score : 0.97
Esto te permitirá comparar modelos rápidamente y decidir cuál funciona mejor en el dataset Iris.
